src/cdk/scrolling/virtual-scroll-viewport.ts

```diff
@@ -0,0 +1,189 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Range} from '@angular/cdk/collections';
+import {CdkForOf} from '@angular/cdk/scrolling/for-of';
+import {
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  DoCheck,
+  ElementRef,
+  Inject,
+  NgZone,
+  OnDestroy,
+  OnInit,
+  ViewChild,
+  ViewEncapsulation,
+  Input,
+} from '@angular/core';
+import {fromEvent} from 'rxjs/observable/fromEvent';
+import {takeUntil} from 'rxjs/operators/takeUntil';
+import {Subject} from 'rxjs/Subject';
+import {Observable} from 'rxjs/Observable';
+import {VIRTUAL_SCROLL_STRATEGY, VirtualScrollStrategy} from './virtual-scroll-strategy';
+
+
+/** A viewport that virtualizes it's scrolling with the help of `CdkForOf`. */
+@Component({
+  moduleId: module.id,
+  selector: 'cdk-virtual-scroll-viewport',
+  templateUrl: 'virtual-scroll-viewport.html',
+  styleUrls: ['virtual-scroll-viewport.css'],
+  host: {
+    'class': 'cdk-virtual-scroll-viewport',
+  },
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  preserveWhitespaces: false,
+})
+export class CdkVirtualScrollViewport implements OnInit, DoCheck, OnDestroy {
+  private _disconnectSubject = new Subject<void>();
+
+  private _renderedRangeSubject = new Subject<Range>();
+
+  /** The direction the viewport scrolls. */
+  @Input() orientation: 'horizontal' | 'vertical' = 'vertical';
+
+  /** The element that wraps the rendered content. */
+  @ViewChild('contentWrapper') _contentWrapper: ElementRef;
+
+  /** The total size of all content, including content that is not currently rendered. */
+  get totalContentSize() { return this._totalContentSize; }
+  set totalContentSize(size: number) {
```

Comment:
These properties shouldn't need to be be getter/setters since they're not `@Input`s for the component (could have an update method, which is more obviously side-effecty)